### DISCUSSION TITLE
Migration guide for 0.22

### DISCUSSION THREAD

**Discussion by rossabaker** [http4s/http4s#6349] [General]
Would be nice to have a summary of the changes for people upgrading from 0.21.

Should include simple instructions for running the scalafix.  The command in https://http4s.org/v0.18/upgrading/ doesn't work for 0.22.

**Comment by rossabaker** [comment]
Scalafixes for 0.22 header changes would be wonderful.  If someone can give some examples of the errors they're hitting even, that could jumpstart the process.

**Comment by rossabaker** [comment]
Demilestoning, because it doesn't block the release.

**Comment by ithinkicancode** [comment]
I was trying out 0.22 the other day. My experience upgrading from 0.21 to 0.22 was pretty smooth; that said, I don't think mine represents the majority of http4s users, because I use Tapir to define my endpoints and Tapir's extension to wire server logic, so I don't really deal with too many http4s API. I do have a docs endpoint that I hand-roll using http4s DSL but that's kinda trivial :)

Anyhow, back to my experience. The changes I needed to make for 0.22 to work so far are:

1. Change `import org.http4s.server.blaze.BlazeServerBuilder` to `import org.http4s.blaze.server.BlazeServerBuilder`
2. Handle the changes in org.http4s.Header. It seems the Header#of is gone and now back to `apply` (think `apply` was deprecated in an older release). I might be wrong though.

That was it, if my memory serves. But again, my experience is probably not common. I hope to learn more from the migration guide. :)

**Comment by kmakiela96** [comment]
I can't use flatMap in for comprehension now because implicit method http4sWithBodySyntax is now gone. How can I transform my request to FlatMapOps?

EDIT: I found it, it is because apply method from Http4sClientDsl in MethodOps class now returns Request[F], not F[Request[F]]. I couldn't find it in changelog so maybe it should be added there.

**Comment by DenisNovac** [comment]
Still can't figure out what to do with `'Content-Type'.renderValue` (this method is gone) and `Method.PermitsBody` (this trait seems to be gone too).

**Comment by armanbilge** [comment]
I'm going to make this into a discussion. 0.22 is EOL as of https://github.com/http4s/http4s/issues/6331 and this is unlikely to land in the docs; at least we can have Q&A.
